src/plan/barriers.rs

```diff
@@ -104,7 +103,10 @@ impl<E: ProcessEdgesWork> ObjectRememberingBarrier<E> {
 
     #[inline(always)]
     fn barrier(&mut self, obj: ObjectReference) {
-        if load_metadata::<E::VM>(&self.meta, obj, None, None) == 0 {
+        // Perform a quick non-atomic load for performance.
+        // It is okay if this check fails occasionally and
+        // the execution goes to the slowpath, we can take care of that in the slowpath.
+        if unsafe { self.meta.load::<E::VM, u8>(obj, None) == 0 } {
```

Comment:
```suggestion
        if self.meta.load_atomic::<E::VM, u8>(obj, None, Ordering::Relaxed) == 0 {
```
The metadata may still be concurrently modified by another thread.  According to C++11 memory model, if this non-atomic load does not have a happens-before relationship with a store, it is still considered a data race, which has undefined behaviour.  So we should still use `load_atomic` if there is any possibility to have a concurrent (no happens-before) store.

If performance is important, we can use the `Ordering::Relaxed` order.  On most platforms, a "relaxed load uses the same instruction as non-atomic load.  But because it uses the Rust-level atomic load, Rust compiler can no longer assume there is no concurrent modifications (because otherwise it is undefined behaviour) and "optimise out" this load.

Reply:
I have made the change as suggested.